```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, fig.width=10, fig.height=7) 
library(tidyverse) 
library(patchwork)  # for combining figures
library(gplots)     # for heatmap.2
library(genefilter) # for calculating row variances
library(RColorBrewer) # color palates
library(Hmisc)     # for pairwise Spearman Correlations
library(ggtree)    # for trees
library(phyloseq)
library(dendextend) # for tanglegrams
library(philentropy) # for JSD
library(Hmisc)  # for rcorr
library(ape)   # for as.phylo
load("gtex_objects.RData")
load("tcr_objects.Rdata")



```


## Incorporating TCR data

```{r}


jsdplist = list()
plot_patient_tcr_tree <- function(df = tcr_pairs, patient="Pt3", metri="JSD", linkage="single"){
  tcr_pairs_matrix <- df %>%
    filter(sample1_clean == patient) %>% 
    filter(sample1_clean == sample2_clean) %>% 
    select(sample1_tissue, sample2_tissue, metri) %>% 
    pivot_wider(id=c(sample1_tissue), names_from = sample2_tissue, values_from=metri, values_fill=0) %>%
    arrange(factor(sample1_tissue, colnames(.)[colnames(.) != "sample1_tissue"])) %>%
    column_to_rownames("sample1_tissue")  %>% 
    as.matrix()
  if(metri == "JSD"){
    hclust_ob <- hclust(as.dist(tcr_pairs_matrix), method = linkage ) %>% as.dendrogram()
    tick_min <- min(1-tcr_pairs_matrix)
  } else{
    hclust_ob <- hclust(1-as.dist(tcr_pairs_matrix), method = linkage)  %>% as.dendrogram()
    tick_min <- min(tcr_pairs_matrix)
  }
  hclust_ob_as_phylo <- ape::as.phylo(hclust_ob)
  hclust_ob_as_phylo$edge.length <- hclust_ob_as_phylo$edge.length*2

  ticks <- make_ticks_df(dendr=hclust_ob  %>% as.dendrogram)
  return(
    ggtree(hclust_ob_as_phylo)  %<+% annotations + xlim(min(ticks$x), 2.2)+#xlim(min(ticks$x), .9)+
      geom_tiplab(aes(color=col, label=tissue_label), 
                  offset = .14, size=2,
                  align = T, linetype = NA, fontface="bold") +
      geom_tippoint() +
      labs(title=patient) + 
      scale_linetype(guide="none") + 
      scale_color_identity()  +
        geom_segment(y=0, yend=0, x=min(ticks$x), xend=max(ticks$x)) + 
        annotate(geom="text",        x=ticks$x, y=-.2, label=ticks$lab, size=2)   + 
        geom_segment(data=ticks, aes(y=-yend, yend=0, x=x, xend=x ))
  )
}
for (pat in unique(tcr_corrs$sample1_clean)){#c("Pt1", "Pt2", "Pt3", "Pt5", "Pt6", "Pt7",  "Pt9")){
  print(pat)
  jsdplist[[pat]] = list(complete=NA, single=NA)
  for (metri in c("JSD")){
    for (li in c("single")){
      if (pat == "Pt3"){
        this_data <- tcr_corrs %>% 
          filter(sample1_tissue !="skin") %>%
          filter(sample2_tissue != "skin")
      } else if(pat== "Pt5"){
        this_data <- tcr_corrs %>% 
          filter(sample1_tissue !="A_marrow") %>%
          filter(sample2_tissue != "A_marrow")
      }else{
        this_data <- tcr_corrs
      }
      p_ <- plot_patient_tcr_tree(df = this_data, 
                                  patient = pat,
                                  linkage=li,
                                  metri=metri) 
      jsdplist[[pat]][[li]] = p_
      ggsave(plot = p_ , filename = file.path(figs_dir, str_glue("TCR_{pat}_{li}_{metri}_tree.pdf")), height=8.5,  width=4.6, )
    }
  }
  #ggsave(plot = wrap_plots(jsdplist[[pat]]), filename = file.path(figs_dir, str_glue("{pat}_trees.pdf")))
}
# stolen from https://stackoverflow.com/a/41882883/4561056
flattenlist <- function(x){  
  morelists <- sapply(x, function(xprime) class(xprime)[1]=="list")
  out <- c(x[!morelists], unlist(x[morelists], recursive=FALSE))
  if(sum(morelists)){ 
    Recall(out)
  }else{
    return(out)
  }
}

ggsave(plot = wrap_plots(flattenlist(lapply(jsdplist, function(x){x[["single"]]})), ncol = 2), filename = file.path(figs_dir, str_glue("TCR_all_pat_trees.pdf")), height = 20, width = 15)
ggsave(plot = wrap_plots(
  list(jsdplist[["Pt1"]][["single"]], 
       jsdplist[["Pt2"]][["single"]],
       jsdplist[["Pt3"]][["single"]],
       jsdplist[["Pt5"]][["single"]],
       jsdplist[["Pt6"]][["single"]],
       jsdplist[["Pt7"]][["single"]],
       jsdplist[["Pt9"]][["single"]]), nrow = 1), filename = file.path(figs_dir, str_glue("TCR_pat_trees_together.pdf")), height = 5, width = 11)
ggsave(plot = wrap_plots(
  list(jsdplist[["C1"]][["single"]], 
       jsdplist[["C2"]][["single"]],
       jsdplist[["C3"]][["single"]]), nrow = 1), filename = file.path(figs_dir, str_glue("TCR_comparator_trees_together.pdf")), height = 5, width = 5.5)


```




## Tanglegrams
```{r plot-tangles, fig.show="hide"}
col_dists <-  jsd_individuals_dist_list[["All"]][[1]]


hclust_col_dists <- jsd_individuals_dist_list[["All"]][[1]] 


dend2 <- jsd_individuals_dist_list[["All"]][[1]]
# labels_colors(dend2) <- data.frame(Tissue = labels(dend2)) %>%
#     left_join(., tissue_key) %>% pull(col)
labels_colors(dend2) <- data.frame(Tissue = labels(dend2)) %>%
    left_join(., tissue_key) %>% pull(col)
labels(dend2) <- data.frame(Tissue = labels(dend2)) %>%
    left_join(., tissue_key)%>% mutate(tmp = gsub("_", "\n", ifelse(is.na(TissueTCR) | TissueTCR == "", Tissue, TissueTCR))) %>% 
  group_by(tmp) %>% mutate(tmp = ifelse(row_number() == 1, tmp, paste0(tmp, row_number()))) %>% pull(tmp) 


plot_patient_tcr_tangle <- function(df = tcr_pairs, patient="Pt3", outdir="figs"){
  df <- df[df$sample1_clean == patient, ] 
  tcr_pairs_matrix <- df %>%
    filter(sample1_clean == sample2_clean) %>% 
    select(sample1_tissue, sample2_tissue, JSD) %>% 
    pivot_wider(id=c(sample1_tissue), names_from = sample2_tissue, values_from=JSD, values_fill=0) %>%
    arrange(factor(sample1_tissue, colnames(.)[colnames(.) != "sample1_tissue"])) %>%
    column_to_rownames("sample1_tissue")  %>% 
    as.matrix()
  
  dend <- hclust(as.dist(tcr_pairs_matrix), method = "single")  %>% as.dendrogram
  

  labels_colors(dend) <- data.frame(TissueTCR = labels(dend)) %>%
    left_join(., tissue_key) %>% pull(col)
  labels(dend) <- gsub("_", "\n", labels(dend))

  #labels_colors(dend2) <- data.frame(TissueTCR = labels(dend2)) %>%
  #  left_join(., tissue_key) %>% pull(col)
  
  entang <- round(entanglement(tanglegram( dend ,  dend2, intersecting = TRUE)), 2)
  dir.create(outdir, showWarnings = FALSE)
  outpath <- file.path(outdir, paste0("tanglegram_", patient, ".png"))
  print(paste0("Writing tanglegram to ", outpath)) 
  png(filename = outpath, width = 8, height = 5, units = "in", res=300)
  ptangle <- tanglegram(
    dend ,  dend2, 
    main_left = "TCR",
    main_right = "GTEx",
    columns_width = c(5,1,5), 
    hang = FALSE, 
    lab.cex =1, intersecting = TRUE,
    margin_inner=5,
    common_subtrees_color_lines = FALSE,
    highlight_distinct_edges  = FALSE,
    highlight_branches_lwd = FALSE, main = paste0("Patient ", patient), sub=paste0( "entanglement = ",entang), cex_main = 2, cex_main_left = 1.5, cex_main_right = 1.5, cex_sub = 1.2)
dev.off()
  return(NA)

}
dir.create(file.path(figs_dir, "tanglegrams"))
for (pat in  unique(tcr_corrs$sample1_clean)){ 
  plot_patient_tcr_tangle(
    df = tcr_corrs,# %>% filter(sample1_tissue !="skin") %>% filter(sample2_tissue != "skin"), 
    patient =pat, outdir=file.path(figs_dir, "tanglegrams"))
}
```


```{r ab-hit-threshold, fig.cap=".", out.width="60%"}
knitr::include_graphics(dir(path = "figs", full.names = TRUE, pattern="tanglegram_Pt.*.png"))
```




```{r}
library(ggrepel)

gtex <- jsd_individuals_dist_list[["All"]][[2]] %>% as.data.frame() %>% rownames_to_column("from") %>% pivot_longer(names_to = "to", cols=-from, values_to = "gtex") %>% 
  mutate(key = paste0(pmin(from, to), pmax(from, to), sep = "")) %>% 
  distinct(key, .keep_all = TRUE) %>% select(-key) %>% 
  inner_join(., tissue_key %>% select(Tissue, TissueTCR), by=c("from"="Tissue")) %>% filter(TissueTCR != "")  %>% 
  select(-from) %>% rename(from="TissueTCR") %>% 
  inner_join(., tissue_key %>% select(Tissue, TissueTCR), by=c("to"="Tissue")) %>% filter(TissueTCR != "")  %>% 
  select(-to) %>% rename(to="TissueTCR")
#   X1 X2 key




tcr <- tcr_corrs %>% 
  rename(from=sample1_tissue, to=sample2_tissue, tcr=JSD, patient=sample1_clean) %>% 
  filter(patient == sample2_clean) %>% 
  filter(from != to) %>% 
  select(from, to, tcr, patient) 

ggsave(
inner_join(gtex, tcr) %>%
  ggplot(aes(color=paste(from, to, sep="-"), y=tcr,x=gtex, label=paste(from, to, sep="-"), shape=patient)) +
  geom_point(size=3) + geom_text_repel(size=2) + 
  scale_color_discrete(guide="none") + 
  scale_shape_manual(values=c(1:10)) + 
  theme(legend.position = "bottom") + 
  labs(x="JSD according to GTEx median summary", y="JSD according to TCR profiles", shape="patient") + theme_classic() ,filename = file.path(figs_dir, "GTEx-vs-TCR-summary.pdf"), width=10, height=10) 
  

```


## Quantifying Anatomic distances with GTEx


```{r}
cluster_order <- data.frame(
  Tissue=labels(jsd_individuals_dist_list[["Tissues_in_TCR"]][[1]])
) %>% right_join(specific_tissue_key)
cluster_order_uniq <- unique(cluster_order$TissueTCR)

corr_df_raw <- tcr %>% 
  filter(patient=="Pt5") %>% 
  filter(to %in% cluster_order$TissueTCR) %>% 
  filter(from %in% cluster_order$TissueTCR) %>% 
  left_join(specific_tissue_key %>% select(BroadTissueTCR, TissueTCR) %>% rename(To=BroadTissueTCR), by=c(to="TissueTCR")) %>% 
  left_join(specific_tissue_key %>% select(BroadTissueTCR, TissueTCR) %>% rename(From=BroadTissueTCR), by=c(from="TissueTCR")) %>%
  left_join(specific_tissue_key %>% select(Tissue, BroadTissueTCR) %>% rename(ToB=Tissue), by=c(To="BroadTissueTCR")) %>% 
  left_join(specific_tissue_key %>% select(Tissue, BroadTissueTCR) %>% rename(FromB=Tissue), by=c(From="BroadTissueTCR")) %>%
  # filter(ToB %in% cluster_order$Tissue) %>% 
  # filter(FromB %in% cluster_order$Tissue) %>% 
#  group_by(to, from) %>%
 # distinct(from, to, .keep_all = TRUE) %>%
  group_by(ToB, FromB) %>%
  dplyr::summarize(tcr=median(tcr)) %>% 
  pivot_wider(names_from="ToB", values_from="tcr") %>% 
  column_to_rownames("FromB") %>% 
 as.matrix() 

corr_df_raw <- corr_df_raw[, rownames(corr_df_raw)]
#corr_df[upper.tri(corr_df)] <- NA
cluster_order <- cluster_order %>% filter(Tissue %in%rownames(corr_df_raw) )
corr_df <- corr_df_raw[cluster_order$Tissue, cluster_order$Tissue]


# pull in the specific tissues GTEx jsd distances, but remove skin
gtex_mat <- jsd_individuals_dist_list[[2]][[2]] 
tissues_in_pt5_tcr <- c("Colon...Sigmoid", "Colon...Transverse", 
                        "Esophagus...Mucosa", 
                        #"Esophagus...Muscularis",
                        "Heart...Left.Ventricle", "Kidney...Cortex", 
                        "Liver", "Skin...Sun.Exposed..Lower.leg.", 
                        "Small.Intestine...Terminal.Ileum", 
                        "Spleen", "Stomach", "Whole.Blood") 
gtex_mat <-gtex_mat[tissues_in_pt5_tcr, tissues_in_pt5_tcr]
hclust_col_dists <- hclust(as.dist(gtex_mat), method = "single") 
dend2 <- hclust_col_dists %>% as.dendrogram
labels_colors(dend2) <- data.frame(Tissue = labels(dend2)) %>%
  left_join(., tissue_key) %>% pull(col)
tmp <- ape::as.phylo(dend2)

tmp$edge.length <- tmp$edge.length*2
ticks <- make_ticks_df(dendr = dend2,dofilter =  TRUE, tick_pos = seq(0, 1, .05)) %>%
  filter(lab < .2)


this_anno <- left_join(annotations, specific_tissue_key %>% select(-col), c("tissue" ="TissueTCR")) %>% select(Tissue, everything())
(p_gtex_spec_tissue <- ggtree(tmp)  %<+% this_anno +
    xlim(min(ticks$x), .4) +
    geom_tiplab(aes(color=col, label=label), 
                offset = 0.01, 
                align = T, linetype = NA, fontface="bold") +
    geom_tippoint() +
    labs() + 
    scale_linetype(guide="none") + 
    scale_color_identity() +
    geom_segment(y=0, yend=0, x=min(ticks$x), xend=max(ticks$x)) + 
    annotate(geom="text",        x=ticks$x, y=-.3, label=ticks$lab )   + 
    geom_segment(data=ticks, aes(y=-yend, yend=0, x=x, xend=x )) +
    theme(legend.position = c(.1, .8))
)


# modified from ggtree to add cell labels 
mygheatmap <- function(p, data, offset=0, width=1, low="green", high="red", color="white",
                     colnames=TRUE, colnames_position="bottom", colnames_angle=0, colnames_level=NULL,label_cells=FALSE,
                     colnames_offset_x = 0, colnames_offset_y = 0, font.size=4, family="", hjust=0.5, legend_title = "value") {

    variable <- value <- lab <- y <- NULL

    ## if (is.null(width)) {
    ##     width <- (p$data$x %>% range %>% diff)/30
    ## }

    ## convert width to width of each cell
    width <- width * (p$data$x %>% range(na.rm=TRUE) %>% diff) / ncol(data)

    isTip <- x <- y <- variable <- value <- from <- to <- NULL

    ## handle the display of heatmap on collapsed nodes
    ## https://github.com/GuangchuangYu/ggtree/issues/242
    ## extract data on leaves (& on collapsed internal nodes) 
    ## (the latter is extracted only when the input data has data on collapsed
    ## internal nodes)
    df <- p$data
    nodeCo <- intersect(df %>% filter(is.na(x)) %>% 
                         select(.data$parent, .data$node) %>% unlist(), 
                     df %>% filter(!is.na(x)) %>% 
                         select(.data$parent, .data$node) %>% unlist())
    labCo <- df %>% filter(.data$node %in% nodeCo) %>% 
        select(.data$label) %>% unlist()
    selCo <- intersect(labCo, rownames(data))
    isSel <- df$label %in% selCo
    
    df <- df[df$isTip | isSel, ]
    start <- max(df$x, na.rm=TRUE) + offset

    dd <- as.data.frame(data)
    ## dd$lab <- rownames(dd)
    i <- order(df$y)

    ## handle collapsed tree
    ## https://github.com/GuangchuangYu/ggtree/issues/137
    i <- i[!is.na(df$y[i])]

    lab <- df$label[i]
    ## dd <- dd[lab, , drop=FALSE]
    ## https://github.com/GuangchuangYu/ggtree/issues/182
    dd <- dd[match(lab, rownames(dd)), , drop = FALSE]


    dd$y <- sort(df$y)
    dd$lab <- lab
    ## dd <- melt(dd, id=c("lab", "y"))
    dd <- gather(dd, variable, value, -c(lab, y))

    i <- which(dd$value == "")
    if (length(i) > 0) {
        dd$value[i] <- NA
    }
    if (is.null(colnames_level)) {
        dd$variable <- factor(dd$variable, levels=colnames(data))
    } else {
        dd$variable <- factor(dd$variable, levels=colnames_level)
    }
    V2 <- start + as.numeric(dd$variable) * width
    mapping <- data.frame(from=dd$variable, to=V2)
    mapping <- unique(mapping)

    dd$x <- V2
    dd$width <- width
    dd[[".panel"]] <- factor("Tree")
    if (is.null(color)) {
        p2 <- p + geom_tile(data=dd, aes(x, y, fill=value, ), width=width, inherit.aes=FALSE)
    } else {
        p2 <- p + geom_tile(data=dd, aes(x, y, fill=value), width=width, color=color, inherit.aes=FALSE)
    }
    if (label_cells){
      p2 <- p2 + geom_text(data=dd, aes(x, y, label=round(value, 2)), color="grey90", size=2)
    }
    if (is(dd$value,"numeric")) {
        p2 <- p2 + scale_fill_gradient(low=low, high=high, na.value=NA, name = legend_title) # "white")
    } else {
        p2 <- p2 + scale_fill_discrete(na.value=NA, name = legend_title) #"white")
    }

    if (colnames) {
        if (colnames_position == "bottom") {
            y <- 0
        } else {
            y <- max(p$data$y) + 1
        }
        mapping$y <- y
        mapping[[".panel"]] <- factor("Tree")
        p2 <- p2 + geom_text(data=mapping, aes(x=to, y = y, label=from), size=font.size, family=family, inherit.aes = FALSE,
                             angle=colnames_angle, nudge_x=colnames_offset_x, nudge_y = colnames_offset_y, hjust=hjust)
    }

    p2 <- p2 + theme(legend.position="right")
    ## p2 <- p2 + guides(fill = guide_legend(override.aes = list(colour = NULL)))

    if (!colnames) {
        ## https://github.com/GuangchuangYu/ggtree/issues/204
        p2 <- p2 + scale_y_continuous(expand = c(0,0))
    }

    attr(p2, "mapping") <- mapping
    return(p2)
}

corr_df_no_tri <- corr_df
corr_df_no_tri[upper.tri(corr_df_no_tri)] <- NA
mygheatmap(p_gtex_spec_tissue, 
         corr_df_no_tri , offset=.3, width=4, label_cells = TRUE,
        colnames=TRUE, legend_title="JSD",colnames_angle=45, hjust=0,
        colnames_offset_y = -.35,
        colnames_position = "top") +
    scale_x_ggtree() + 
    scale_y_continuous(expand=c(0, 0.4)) + scale_fill_viridis_c(na.value = "white", limits=c(0,1)) +
  labs(fill="JSD") + theme(axis.text = element_text(size=3)) + ylim(c(-1, 16)) + xlim(-.09, 1.04) 
  

ggsave(p_gtex_spec_tissue, filename = file.path(figs_dir, "GTEx_Pt5_tissues.pdf"), height=5, width=4)

```


# Inset
```{r}

dist_list <- jsd_individuals_dist_list
color_min <- min(sapply(dist_list, function(x) {min(x[[2]])}))
color_max <- max(sapply(dist_list, function(x) {max(x[[2]])}))

tmp <- ape::as.phylo(jsd_individuals_dist_list[["Tissues_in_TCR"]][[1]])

tmp$edge.length <- tmp$edge.length*2
ticks <- make_ticks_df(dendr = jsd_individuals_dist_list[["Tissues_in_TCR"]][[1]],dofilter =  TRUE, tick_pos = seq(0,1,.1))

# see  https://stackoverflow.com/questions/58492013/possible-to-force-non-occurring-elements-to-show-in-ggplot-legend
# for why we use "limits" not "breaks" in the manual color scale
(p <- ggtree(tmp)  %<+% specific_tissue_key + xlim(min(ticks$x), .9) +
    geom_tiplab(aes(color=TissueSystem), 
                offset = 0.01,#(1-max(tmp$edge.length)) + .01, 
                align = T, linetype = NA) +
    geom_tippoint() +
    #geom_vline(xintercept = min(ticks$x))+
    labs(title="Tissues in TCR") + 
    scale_linetype(guide="none") + 
    scale_color_manual(
      values = specific_tissue_key$col,
      limits = specific_tissue_key$TissueSystem,
      labels = specific_tissue_key$TissueSystem) +
    geom_segment(y=0, yend=0, x=min(ticks$x), xend=max(ticks$x)) + 
    annotate(geom="text",        x=ticks$x, y=-.2, label=ticks$lab )   + 
    geom_segment(data=ticks, aes(y=-yend, yend=0, x=x, xend=x ))
)
pmat<- ggplot(dist_list[[i]][[2]] %>%
                as.data.frame() %>% 
                rownames_to_column("x") %>% 
                pivot_longer(names_to="y", values_to = "val", -x), aes(x, y, fill=val)) + 
  geom_tile() + 
  coord_fixed(ratio=1) + 
  geom_text(aes(label=round(val, 2)), color="grey50", size=4)+
  scale_fill_viridis_c(option = "A", breaks=seq(0, 1, .2), limits=c(color_min, color_max)) +
  guides(fill = guide_colourbar(direction = "horizontal")) + 
  scale_y_discrete(position = "right", expand = c(0,0)) +
  scale_x_discrete( expand = c(0,0)) +
  labs(subtitle=corrtype) +
  theme(axis.text.x = element_text(angle=45, hjust=1, size=9))
plist[[corrtype]][[i]] <- p
plistheat[[corrtype]][[i]] <- p +   pmat
```

## GTEx Methods
GTEx data for version 8 was retrieved from https://www.gtexportal.org/home/datasets/ on October 12, 2021. We filtered the expression data similar to (https://www.ncbi.nlm.nih.gov/labs/pmc/articles/PMC4547472/) as follows. Protein-coding genes were identified from the Ensembl 104 by filtering the annotations to retain "protein_coding" gene_biotypes with "gene" types, according to Homo_sapiens.GRCh38.104.chr.gtf. Mitochondrial, hemoglobin ("HBB"), and MTATP6P1 (https://www.nature.com/articles/s41598-018-23226-4) transcripts were removed. Genes with low expression (<2 TPM) were removed, and genes with high expression were capped at 10K TMP. Genes fold change < 2 (relative to sample with the lowest expression, per-gene) and an absolute difference of less than 10 TPM were removed. The same filtering was applied to the median-aggregated and the per-individual expression data.


(optional, as I don't think any of these analyses made it in to the paper) For per-individual analyses from the GTEx data, we applied the following filtering criteria.  For non-kidney tissue, we required the RIN score to be above 7 and an autolysis score of zero (when available); we did not require this for kidney tissue as no samples met this threshhold.  For consistency we selected samples wiht a read length of 76, and limited the analysis to those samples included in the GTEx Analysis Freeze.  Lastly, we retained individuals who had at least 5 tissues available.   

JSD was used to quantify the between-tissue similarity, and was plotted as dendrograms as described above, using ggtree(https://doi.org/10.1111/2041-210X.12628 , https://academic.oup.com/mbe/article/35/12/3041/5142656,  https://doi.org/10.1002/cpbi.96), dendextend (https://academic.oup.com/bioinformatics/article/31/22/3718/240978), and ape (https://academic.oup.com/bioinformatics/article/35/3/526/5055127?login=true). See supplementary repository for a full list of all packages and versions used.



